### DISCUSSION TITLE
Question: Automatically Update vs Allows Automatic Updates

### DISCUSSION THREAD

**Discussion by dsmurfin** [sparkle-project/Sparkle#2281] [Q&A] [answered]
I am trying to provide an app settings panel for configuring updates.

There are a couple of Info.plist settings which seem to have overlapping functionality, and the documentation isn't as clear as maybe it could be.

`SUAllowsAutomaticUpdates`
By default, Sparkle automatically presents your users with the option to allow to automatically download and install any available updates if automatic checking of updates is enabled.Set this to NO to disallow automatic updates and require manual installation every time.Set this to YES to always allow automatic updates even if automatic checking of updates is disabled.

`SUAutomaticallyUpdate`
Default: NO. Enables automatic download and installation of updates. If set to YES, Sparkle will attempt to download and install new updates silently in the background. In Sparkle 1, updates won’t be opted into this if users need to provide authorization. In Sparkle 2, updates will be downloaded but not installed automatically if authorization is required. In all versions of Sparkle, if the application hasn’t quit for 1 week the user will be presented with installing the downloaded update (unless the updater delegate overrides this).

Would someone be able to explain the difference between these two flags?

On a related note, if I add `SUEnableAutomaticChecks = true` to Info.plist Sparkle also adds `SUAutomaticallyUpdate = true` to NSUSerDefaults at first launch. These two options feel independent. Why does enabling checks also enable this second option?

Thanks so much.

**Comment by zorgiepoo** [comment, marked as answer]
With `SUAllowsAutomaticUpdates` this controls whether or not Sparkle is allowed to show an option to the user that allows them to turn on/off automatic downloading/installing of updates. You usually don't need to configure this, especially if you want to allow your users to opt into automatic downloading/installing and let Sparkle schedule checking for updates automatically.

`SUAutomaticallyUpdate` specifies by default if Sparkle should try to automatically download and install updates when Sparkle checks for updates (default is NO). Otherwise Sparkle may always present the update to the user when it checks for updates in the background.

`SUEnableAutomaticChecks` specifies if Sparkle schedules and performs checks for updates automatically. You don't have to set this. By default users will be prompted for permission for Sparkle to check for updates automatically on second launch.

> On a related note, if I add SUEnableAutomaticChecks = true to Info.plist Sparkle also adds SUAutomaticallyUpdate = true to NSUSerDefaults at first launch.

No. This may be your own doing. Sparkle (at least 2) does not set `-[SPUUpdater setAutomaticallyDownloadsUpdates:]` by itself (unless the user changes it). This page has some API expectations that may clarify some things as well: https://sparkle-project.org/documentation/programmatic-setup/#api-expectations -- and it links to samples for adding settings.

**Comment by dsmurfin** [comment]
Thank you, this is really helpful.

I am hoping to be able to provide some more granular user facing options along the lines of what is indicated below, so understanding each of these keys is quite important. 

I think what I might be looking to do would work by settings `SUAllowsAutomaticUpdates = NO`, and using the appropriate API to set these keys `setAutomaticallyDownloadsUpdates` etc.

Manual - `SUEnableAutomaticChecks = NO` + `SUAutomaticallyUpdate = NO`
Check for updates - `SUEnableAutomaticChecks = YES` + `SUAutomaticallyUpdate = NO`
Download and install updates - `SUEnableAutomaticChecks = YES` + `SUAutomaticallyUpdate = YES`

> No. This may be your own doing. Sparkle (at least 2) does not set -[SPUUpdater setAutomaticallyDownloadsUpdates:] by itself (unless the user changes it). This page has some API expectations that may clarify some things as well: https://sparkle-project.org/documentation/programmatic-setup/#api-expectations -- and it links to samples for adding settings.

Thanks for the clarification. I did try clearing user defaults a few times, and launching again, so I will have to check again.

Appreciate the detailed response.

**Reply by zorgiepoo** [reply]
> I think what I might be looking to do would work by settings SUAllowsAutomaticUpdates = NO, and using the appropriate API to set these keys setAutomaticallyDownloadsUpdates etc.

If you set `SUAllowsAutomaticUpdates` to NO, then it does not make sense to call `setAutomaticallyDownloadsUpdates` later. Automatic downloading of updates is just disabled. You only set it to NO if you don't want to allow automatic downloading/installing of updates.

---

Do you want Sparkle to ask the user's consent/permission for checking for updates automatically (on second launch), which will then set `automaticallyChecksForUpdates` based on the user's response? If so, don't set `SUEnableAutomaticChecks`.

Do you want to make it not possible for the app to download/install updates automatically? If so, set `SUAllowsAutomaticUpdates` to NO.

Do you want to enable downloading/installing of updates automatically by default? If so, set `SUAutomaticallyUpdate` to YES.

Do you want to not let Sparkle automatically schedule checking for updates, and potentially handle scheduling update checks yourself? If so, set  `SUEnableAutomaticChecks` to NO. If you want to allow automatic downloading/installing of updates in this case, set `SUAllowsAutomaticUpdates` to YES.

As you can see, the default (not setting anything) is Sparkle asks user's consent to check for updates automatically + downloading/installing of updates automatically is allowed + downloading/installing of updates automatically is initially off.

**Reply by dsmurfin** [reply]
Thank you, I think I've finally got it.

I was confused from the previous answer and thought `SUAllowsAutomaticUpdates` was only related to whether the checkbox option was included in the dialog, not that it actually impacted functionality.

Thank for the clarification.
